core/include/vecmem/containers/device_vector_data.ipp

```diff
@@ -0,0 +1,65 @@
+/** VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+namespace vecmem {
+
+   template< typename TYPE >
+   VECMEM_HOST_AND_DEVICE
+   device_vector_data< TYPE >::
+   device_vector_data( std::size_t size, pointer ptr )
+   : m_size( size ), m_ptr( ptr ) {
+
+   }
+
+   template< typename TYPE >
+   template< typename ALLOC >
+   VECMEM_HOST
+   device_vector_data< TYPE >::
+   device_vector_data( std::vector< value_type, ALLOC >& vec )
+   : m_size( vec.size() ), m_ptr( vec.data() ) {
+
+   }
+
+   template< typename TYPE >
+   template< typename ALLOC >
+   VECMEM_HOST
+   device_vector_data< TYPE >::
+   device_vector_data( const std::vector< value_type, ALLOC >& vec )
+   : m_size( vec.size() ), m_ptr( vec.data() ) {
+
+   }
+
+   template< typename TYPE >
+   template< typename OTHERTYPE,
+             std::enable_if_t< ( ! std::is_same< TYPE, OTHERTYPE >::value ) &&
+                               std::is_convertible< OTHERTYPE*, TYPE* >::value,
+                               bool > >
+   VECMEM_HOST_AND_DEVICE
+   device_vector_data< TYPE >::
+   device_vector_data( const device_vector_data< OTHERTYPE >& parent )
+   : m_size( parent.m_size ), m_ptr( parent.m_ptr ) {
```

Comment:
Are you sure this works? What if `OTHERTYPE` is a larger child class of `TYPE`, then the pointer arithmetic would no longer add up, would it?

Reply:
I didn't run too rigorous tests, but I think automatic pointer conversion really only works for going from `FOO*` to `const FOO*`. Everything else needs to be explicitly declared as a conversion, no? :confused:

Reply:
But we're throwing away the original type of the vector without actually converting anything on a bit-level. So I would think it possible for the array pointed at by `m_ptr` to consist of objects of a different size from `TYPE`.

Reply:
No, this does seem to be reasonably safe. You should try the following:
  - This does compile:

```c++
vecmem::vector< int > hostvec;
vecmem::device_vector_data< const int > test( hostvec );
```

  - But the following fails miserably:

```c++
vecmem::vector< int > hostvec;
vecmem::device_vector_data< short > test( hostvec );
```

Like:

```
/home/krasznaa/ATLAS/projects/vecmem/vecmem/tests/core/test_core_containers.cpp:56:40: error: no matching constructor for initialization of 'vecmem::device_vector_data<short>'
   vecmem::device_vector_data< short > dummy( m_reference_vector );
                                       ^      ~~~~~~~~~~~~~~~~~~
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:26:11: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'vecmem::vector<int>' (aka 'vector<int, polymorphic_allocator<int>>') to 'const vecmem::device_vector_data<short>' for 1st argument
   struct device_vector_data {
          ^
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:26:11: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'vecmem::vector<int>' (aka 'vector<int, polymorphic_allocator<int>>') to 'vecmem::device_vector_data<short>' for 1st argument
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:42:7: note: candidate template ignored: could not match 'short' against 'int'
      device_vector_data( std::vector< value_type, ALLOC >& vec );
      ^
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:46:7: note: candidate template ignored: could not match 'short' against 'int'
      device_vector_data( const std::vector< value_type, ALLOC >& vec );
      ^
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:60:7: note: candidate template ignored: could not match 'device_vector_data' against 'vector'
      device_vector_data( const device_vector_data< OTHERTYPE >& parent );
      ^
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:34:7: note: candidate constructor not viable: requires 0 arguments, but 1 was provided
      device_vector_data() = default;
      ^
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:37:7: note: candidate constructor not viable: requires 2 arguments, but 1 was provided
      device_vector_data( std::size_t size, pointer ptr );
      ^
1 error generated.
```

Reply:
But what if you do:

```cpp
vecmem::vector<int> hostvec;
vecmem::device_vector_data<long> test(hostvec);
```

Reply:
```
/home/krasznaa/ATLAS/projects/vecmem/vecmem/tests/core/test_core_containers.cpp:56:39: error: no matching constructor for initialization of 'vecmem::device_vector_data<long>'
   vecmem::device_vector_data< long > dummy( m_reference_vector );
                                      ^      ~~~~~~~~~~~~~~~~~~
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:26:11: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'vecmem::vector<int>' (aka 'vector<int, polymorphic_allocator<int>>') to 'const vecmem::device_vector_data<long>' for 1st argument
   struct device_vector_data {
          ^
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:26:11: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'vecmem::vector<int>' (aka 'vector<int, polymorphic_allocator<int>>') to 'vecmem::device_vector_data<long>' for 1st argument
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:42:7: note: candidate template ignored: could not match 'long' against 'int'
      device_vector_data( std::vector< value_type, ALLOC >& vec );
      ^
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:46:7: note: candidate template ignored: could not match 'long' against 'int'
      device_vector_data( const std::vector< value_type, ALLOC >& vec );
      ^
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:60:7: note: candidate template ignored: could not match 'device_vector_data' against 'vector'
      device_vector_data( const device_vector_data< OTHERTYPE >& parent );
      ^
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:34:7: note: candidate constructor not viable: requires 0 arguments, but 1 was provided
      device_vector_data() = default;
      ^
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/device_vector_data.hpp:37:7: note: candidate constructor not viable: requires 2 arguments, but 1 was provided
      device_vector_data( std::size_t size, pointer ptr );
      ^
1 error generated.
```

Pointers **do not convert easily**...

Reply:
Okay, hold on, this needs to become a little more complex:

```cpp
class A {
    int m_a;
};

class B : public A {
    int m_b;
};

vecmem::vector<B> hostvec;
vecmem::device_vector_data<A> test(hostvec);
```

Does that also throw an error?

Reply:
No, that probably does not.

Fine, let's declare this constructor more narrowly. The code does need to be able to convert from `vecmem::device_vector_data<FOO>` to `vecmem::device_vector_data<const FOO>`. What's your proposal, what should the SFINAE expression be for only enabling that sort of `OTHERTYPE`?

Reply:
Something like this, perhaps?

```cpp
template<typename OTHERTYPE,
    std::enable_if_t<
        (!std::is_same<TYPE, OTHERTYPE>::value) &&
        std::is_same<TYPE, std::add_const<OTHER_TYPE>::type>::value,
        bool
    >
>
```

Reply:
:thinking: Okay, that's probably the way to go... I was thinking of using [std::remove_cv](https://en.cppreference.com/w/cpp/types/remove_cv) in some way, but that would still allow for some invalid combinations...